crates/codegen/parser/src/code_generator.rs

```diff
@@ -3,47 +3,121 @@ use std::{
     path::PathBuf,
 };
 
+use codegen_schema::types::grammar::Grammar;
 use codegen_utils::context::CodegenContext;
 use inflector::Inflector;
 use proc_macro2::{Ident, TokenStream};
 use quote::{format_ident, quote};
 use semver::Version;
 
-#[derive(Clone, Debug, Default)]
-pub struct VersionedFunctionBody {
-    pub comment: String,
-    pub versions: BTreeMap<Version, TokenStream>,
-}
+#[derive(Clone, Debug)]
+pub struct VersionedFunctionBody(BTreeMap<Version, Option<(String, TokenStream)>>);
 
 impl VersionedFunctionBody {
-    fn insert(&mut self, comment: &str, version: &Version, body: TokenStream) {
-        self.comment.push_str(comment);
-        self.versions.insert(version.clone(), body);
+    fn new(version: &Version) -> Self {
```

Comment:
I wonder if `is_defined_for_all_versions` should check that the first defined version is the first version of the grammar? in case a production was introduced later.
Otherwise, if `VersionedFunctionBody::new()` is always initialized with the first grammar version, thoughts on renaming it to indicate such?

```suggestion
    fn new(first_grammar_version: &Version) -> Self {
```

Reply:
1. Note that this constructor inserts a `None` for the first version, so that is already handled.
2. done